[css-anchor-position-1] Trigger evaluation of default anchor when it's used for anchor-center

anttijk opened the pull request:
#### aa6b7858fefa58b92a4ba4da1df0b837fb5cf1c6
<pre>
[css-anchor-position-1] Trigger evaluation of default anchor when it&apos;s used for anchor-center
<a href="https://bugs.webkit.org/show_bug.cgi?id=283295">https://bugs.webkit.org/show_bug.cgi?id=283295</a>
<a href="https://rdar.apple.com/problem/140124449">rdar://problem/140124449</a>

Reviewed by Alan Baradlay.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-offset-change-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-013-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-014-expected.txt:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::findAnchorForAnchorFunctionAndAttemptResolution):

Track in state whether the we are using anchor() or anchor-size().

(WebCore::Style::AnchorPositionEvaluator::evaluate):
(WebCore::Style::AnchorPositionEvaluator::evaluateSize):
(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositioningStatesAfterInterleavedLayout):

Invalidate layout if after resolving anchors if using layout-time anchor positioning.

(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositionedStateForLayoutTimePositioned):

Layout time anchor positioning requires state too. Create it if neeeded.

(WebCore::Style::AnchorPositionEvaluator::isLayoutTimeAnchorPositioned):

Helper to test the style to see if we are using layout-time anchor positioning.

(WebCore::Style::AnchorPositionEvaluator::findAnchorAndAttemptResolution): Deleted.
* Source/WebCore/style/AnchorPositionEvaluator.h:

Rename some enum values for clarity.

* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::invalidateForAnchorDependencies):

Invalidate layout (instead of style) if we are using layout-time anchoring.

(WebCore::Style::Scope::resetAnchorPositioningStateBeforeStyleResolution):
(WebCore::Style::Scope::updateAnchorPositioningStateAfterStyleResolution):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::updateAnchorPositioningState):

Create state for layout-time anchor positioning after style resolution if needed.

Canonical link: <a href="https://commits.webkit.org/290534@main">https://commits.webkit.org/290534@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d664d858ebc6358c80528673e8d1942b05762c71

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90372 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9902 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45299 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95378 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41148 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92424 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10289 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18220 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/69570 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27152 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93373 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/7888 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81987 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49927 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/7615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/36351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40280 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77941 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37420 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97210 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17560 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12926 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/78571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17817 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77810 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/77792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19200 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22231 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10827 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17570 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17311 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20763 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19095 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tuankiet65 commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -412,15 +414,15 @@ RefPtr<Element> AnchorPositionEvaluator::findAnchorAndAttemptResolution(const Bu
         // If anchor resolution has progressed past Initial, and we pick up a new anchor name, set the
         // stage back to Initial. This restarts the resolution process to resolve newly added names.
```
Change the enum name in the comment too (or rewrite it)

tuankiet65 commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -734,18 +736,33 @@ void AnchorPositionEvaluator::updateAnchorPositioningStatesAfterInterleavedLayou
                 state.anchorElements = findAnchorsForAnchorPositionedElement(element, state.anchorNames, anchorsForAnchorName);
-            state.stage = AnchorPositionResolutionStage::FoundAnchors;
+                if (isLayoutTimeAnchorPositioned(renderer->style()))
+                    renderer->setNeedsLayout();
```
Maybe I'm missing something, but why do we invalidate layout here too when we've already invalidated in `Scope::invalidateForAnchorDependencies`?

anttijk replied:
`invalidateForAnchorDependencies` is only triggered when something changes in the layout. In case of the first resolution, the anchor and anchored will be positioned (by the first interleaved layout) before we have connected the anchors so we need another layout for the anchored elements. Basically needed to get the first PASS in imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-offset-change.html

Not super pretty but works for now.

anttijk committed:
[css-anchor-position-1] Trigger evaluation of default anchor when it's used for anchor-center
https://bugs.webkit.org/show_bug.cgi?id=283295
rdar://problem/140124449

Reviewed by Alan Baradlay.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-offset-change-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-013-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-014-expected.txt:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::findAnchorForAnchorFunctionAndAttemptResolution):

Track in state whether the we are using anchor() or anchor-size().

(WebCore::Style::AnchorPositionEvaluator::evaluate):
(WebCore::Style::AnchorPositionEvaluator::evaluateSize):
(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositioningStatesAfterInterleavedLayout):

Invalidate layout if after resolving anchors if using layout-time anchor positioning.

(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositionedStateForLayoutTimePositioned):

Layout time anchor positioning requires state too. Create it if neeeded.

(WebCore::Style::AnchorPositionEvaluator::isLayoutTimeAnchorPositioned):

Helper to test the style to see if we are using layout-time anchor positioning.

(WebCore::Style::AnchorPositionEvaluator::findAnchorAndAttemptResolution): Deleted.
* Source/WebCore/style/AnchorPositionEvaluator.h:

Rename some enum values for clarity.

* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::invalidateForAnchorDependencies):

Invalidate layout (instead of style) if we are using layout-time anchoring.

(WebCore::Style::Scope::resetAnchorPositioningStateBeforeStyleResolution):
(WebCore::Style::Scope::updateAnchorPositioningStateAfterStyleResolution):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::updateAnchorPositioningState):

Create state for layout-time anchor positioning after style resolution if needed.

Canonical link: https://commits.webkit.org/290534@main
